### PR TITLE
Fix ignored tests in the desired-balance feature branch

### PR DESCRIPTION
**Pull request by idegtiarenko**

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java [hunk 1]
```diff
@@ -383,7 +383,6 @@ public void testRestoreIndexWithMissingShards() throws Exception {
         assertDocCount("test-idx-closed", 100L);
     }
 
-    @AwaitsFix(bugUrl = "TODO")
     public void testRestoreIndexWithShardsMissingInLocalGateway() throws Exception {
```
Comment (idegtiarenko) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java:386]: Executed this 250 times. No failures so far

File: server/src/internalClusterTest/java/org/elasticsearch/recovery/FullRollingRestartIT.java [hunk 2]
```diff
@@ -170,7 +170,6 @@ public void testFullRollingRestart() throws Exception {
         }
     }
 
-    @AwaitsFix(bugUrl = "TODO")
     public void testNoRebalanceOnRollingRestart() throws Exception {
```
Comment (idegtiarenko) [inline server/src/internalClusterTest/java/org/elasticsearch/recovery/FullRollingRestartIT.java:173]: This test was broken as we were not detecting delayed shards as ignored

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceInput.java [hunk 3]
```diff
@@ -37,10 +38,10 @@ private static Set<ShardRouting> getIgnoredShardsWithDiscardedAllocationStatus(R
...
             .stream()
-            .map(
-                shardRouting -> shardRouting.updateUnassigned(
-                    discardAllocationStatus(shardRouting.unassignedInfo()),
-                    shardRouting.recoverySource()
+            .flatMap(
+                shardRouting -> Stream.of(
+                    shardRouting,
+                    shardRouting.updateUnassigned(discardAllocationStatus(shardRouting.unassignedInfo()), shardRouting.recoverySource())
                 )
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceInput.java:45]: Related to https://github.com/elastic/elasticsearch/pull/90600

In some cases the allocation status is coming from GatewayAllocator and need to be reseted to `NO_ATTEMPT` to be able to find the ignored shard in others (NO_VALID_SHARD_COPY) it is kept as is.

I wonder if this indicates that we want to have some other way of identifying ignored shards bypassing its entire state and unassigned info?

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**